File: backend/src/modules/transportation/deductions/deductions.service.ts
import { Injectable } from '@nestjs/common';
import * as ExcelJS from 'exceljs';
import { PrismaService } from '../../../prisma/prisma.service';
import { fail, paginationHeader, success, successList, successWrite } from '../../../common/response/base-response';
import { DeductionBody } from './deductions.controller';

// Rich include so list rows can be flattened to the frozen JSON keys.
const deductionInclude = {
  supplier: true,
  route: { include: { contactPerson: true } },
} as const;

export interface DeductionFilters {
  supplierId?: number;
  routeId?: number;
  fromDate?: string;
  toDate?: string;
  name?: string; // route name search
}

@Injectable()
export class DeductionsService {
  constructor(private readonly prisma: PrismaService) {}

  /** Resolve a set of creator (User) ids → their full names, in one query. */
  private async creatorNames(ids: (number | null | undefined)[]): Promise<Map<number, string>> {
    const unique = [...new Set(ids.filter((id): id is number => !!id))];
    const map = new Map<number, string>();
    if (unique.length === 0) return map;
    const users = await this.prisma.user.findMany({
      where: { id: { in: unique } },
      select: { id: true, firstName: true, middleName: true, lastName: true, email: true },
    });
    for (const u of users) {
      const name = [u.firstName, u.middleName, u.lastName].filter(Boolean).join(' ') || u.email;
      map.set(u.id, name);
    }
    return map;
  }

  /** Shared WHERE builder for the list + Excel export (same filters). */
  private buildWhere(filters: DeductionFilters): Record<string, unknown> {
    const where: Record<string, unknown> = {};
    if (filters.supplierId) where.supplierId = filters.supplierId;
    if (filters.routeId) where.routeId = filters.routeId;
    if (filters.name) where.route = { is: { nameOfRoute: { contains: filters.name } } };
    if (filters.fromDate || filters.toDate) {
      const range: Record<string, Date> = {};
      if (filters.fromDate) range.gte = new Date(filters.fromDate);
      if (filters.toDate) range.lte = new Date(filters.toDate);
      where.dateOfDeduction = range;
    }
    return where;
  }

  /** GET getAllDeduction — paginated deductions (TransportationDeductionData rows). */
  async getAll(pageNo = 1, noOfItems = 20, filters: DeductionFilters = {}) {
    const where = this.buildWhere(filters);

    const [total, rows] = await this.prisma.$transaction([
      this.prisma.transportationVehicleRouteDeduction.count({ where }),
      this.prisma.transportationVehicleRouteDeduction.findMany({
        where,
        include: deductionInclude,
        orderBy: { id: 'desc' },
        skip: (pageNo - 1) * noOfItems,
        take: noOfItems,
      }),
    ]);

    const creators = await this.creatorNames(rows.map((d) => d.creationBy));

    const data = rows.map((d) => ({
      Id: d.id,
      SupplierName: d.supplier?.name ?? '',
      SupplierId: d.supplierId ?? 0,
      TransportationVehicleRouteName: d.route?.nameOfRoute ?? '',
      TransportationVehicleRouteId: d.routeId,
      Serial: d.serial ?? '0',
      Date: d.dateOfDeduction ? d.dateOfDeduction.toISOString() : '',
      CreationDate: d.creationDate ? d.creationDate.toISOString() : '',
      DeductPerRound: d.deductPerRound ?? 0,
      Cause: d.cause ?? '',
      CreationName: creators.get(d.creationBy ?? -1) ?? '', // creator (User) full name
      CreationBy: d.creationBy ?? 0,
      SupplierContentPersonIdName: d.route?.contactPerson?.name ?? '', // driver name (frozen typo key: Content vs Contact)
      FromDate: d.fromDate ? d.fromDate.toISOString() : '',
      ToDate: d.toDate ? d.toDate.toISOString() : '',
      typeOfDeduction: d.typeOfDedct ?? '', // lowercase frozen key
      // Faithful to the old getAllDeduction: routePrice is the route's CURRENT
      // LineCost (dynamic), not a stored deduction column. (old svc:2781)
      routePrice: d.route?.lineCost ?? d.routePrice ?? 0, // lowercase frozen key
    }));

    return successList(data, paginationHeader(pageNo, noOfItems, total));
  }

  /** GET getAllDeductionExcel — the filtered deductions as a base64 (.xlsx) file. */
  async excel(filters: DeductionFilters = {}) {
    const rows = await this.prisma.transportationVehicleRouteDeduction.findMany({
      where: this.buildWhere(filters),
      include: deductionInclude,
      orderBy: { id: 'desc' },
    });

    const workbook = new ExcelJS.Workbook();
    const sheet = workbook.addWorksheet('الخصومات');
    sheet.views = [{ rightToLeft: true }];
    sheet.columns = [
      { header: 'الخط', key: 'route', width: 26 },
      { header: 'المورد', key: 'supplier', width: 22 },
      { header: 'السيريال', key: 'serial', width: 14 },
      { header: 'التاريخ', key: 'date', width: 14 },
      { header: 'قيمة الخصم', key: 'amount', width: 14 },
      { header: 'النوع', key: 'type', width: 12 },
      { header: 'السبب', key: 'reason', width: 30 },
    ];
    sheet.getRow(1).font = { bold: true };
    for (const d of rows) {
      sheet.addRow({
        route: d.route?.nameOfRoute ?? '',
        supplier: d.supplier?.name ?? '',
        serial: d.serial ?? '',
        date: d.dateOfDeduction ? d.dateOfDeduction.toISOString().slice(0, 10) : '',
        amount: d.deductPerRound ?? 0,
        type: (d.typeOfDedct ?? '').toLowerCase() === 'taxes' ? 'ضرائب' : 'عادي',
        reason: d.cause ?? '',
      });
    }

    const buffer = await workbook.xlsx.writeBuffer();
    return success(Buffer.from(buffer).toString('base64'));
  }

  /** POST AddDeduction — create a deduction. creationBy = caller. */
  async add(body: DeductionBody, userId: number) {
    const routeId = Number(body?.TransportationVehicleRouteId) || 0;
    if (!routeId) return fail('Err101', 'TransportationVehicleRouteId is required');

    const created = await this.prisma.transportationVehicleRouteDeduction.create({
      data: { ...this.toData(body, routeId), creationBy: userId },
    });
    return successWrite(created.id);
  }

  /** POST UpdateDeduction — update a deduction. Preserves the original creator
   *  (old UpdateDeduction never touches CreationBy — svc:7277-7283). */
  async update(id: number, body: DeductionBody) {
    if (!id) return fail('Err101', 'Id is required');
    const routeId = Number(body?.TransportationVehicleRouteId) || 0;
    await this.prisma.transportationVehicleRouteDeduction.update({
      where: { id },
      data: this.toData(body, routeId),
    });
    return successWrite(id);
  }

  /** Maps the documented payload to the deduction model columns (no creator —
   *  the caller adds creationBy only on create). */
  private toData(body: DeductionBody, routeId: number) {
    return {
      supplierId: body.SupplierId ? Number(body.SupplierId) : null,
      routeId,
      serial: body.Serial ?? '',
      dateOfDeduction: body.DateOfDeduction ? new Date(body.DateOfDeduction) : new Date(),
      deductPerRound: Number(body.DeductPerRound) || 0,
      cause: body.Cause ?? null,
      typeOfDedct: body.TypeOfDedct ?? 'Normal',
    };
  }
}

File: backend-dotnet/GarasTransport.Api/Modules/Transportation/Deductions/DeductionsService.cs
using ClosedXML.Excel;
using GarasTransport.Api.Common;
using Microsoft.EntityFrameworkCore;

namespace GarasTransport.Api.Modules.Transportation.Deductions;

// Shape of the Add/Update deduction payload (documented PascalCase keys, §6.10).
public class DeductionBody
{
    public int? Id { get; set; }
    public int? SupplierId { get; set; }
    public string? Serial { get; set; }
    public int? TransportationVehicleRouteId { get; set; }
    public double? DeductPerRound { get; set; }
    public string? Cause { get; set; }
    public string? DateOfDeduction { get; set; }
    public string? TypeOfDedct { get; set; } // Taxes | Normal
}

public class DeductionFilters
{
    public int? SupplierId { get; set; }
    public int? RouteId { get; set; }
    public string? FromDate { get; set; }
    public string? ToDate { get; set; }
    public string? Name { get; set; } // route name search
}

public class DeductionsService
{
    private readonly AppDbContext _db;

    public DeductionsService(AppDbContext db) => _db = db;

    /// <summary>Resolve a set of creator (User) ids → their full names, in one query.</summary>
    private async Task<Dictionary<int, string>> CreatorNames(IEnumerable<int?> ids)
    {
        var unique = ids.Where(i => i.HasValue).Select(i => i!.Value).Distinct().ToList();
        var map = new Dictionary<int, string>();
        if (unique.Count == 0) return map;
        var users = await _db.Users.Where(u => unique.Contains(u.Id)).ToListAsync();
        foreach (var u in users)
        {
            var name = string.Join(' ', new[] { u.FirstName, u.MiddleName, u.LastName }.Where(s => !string.IsNullOrEmpty(s)));
            map[u.Id] = string.IsNullOrEmpty(name) ? u.Email : name;
        }
        return map;
    }

    private IQueryable<TransportationVehicleRouteDeduction> BuildQuery(DeductionFilters filters)
    {
        var q = _db.TransportationVehicleRouteDeductions.AsQueryable();
        if (filters.SupplierId.HasValue) q = q.Where(d => d.SupplierId == filters.SupplierId.Value);
        if (filters.RouteId.HasValue) q = q.Where(d => d.RouteId == filters.RouteId.Value);
        if (!string.IsNullOrEmpty(filters.Name)) q = q.Where(d => d.Route!.NameOfRoute.Contains(filters.Name));
        var from = ParseDate(filters.FromDate);
        var to = ParseDate(filters.ToDate);
        if (from.HasValue) q = q.Where(d => d.DateOfDeduction >= from.Value);
        if (to.HasValue) q = q.Where(d => d.DateOfDeduction <= to.Value);
        return q;
    }

    /// <summary>GET getAllDeduction — paginated deductions (TransportationDeductionData rows).</summary>
    public async Task<object> GetAll(int pageNo, int noOfItems, DeductionFilters filters)
    {
        var query = BuildQuery(filters);
        var total = await query.CountAsync();
        var rows = await query
            .Include(d => d.Supplier)
            .Include(d => d.Route).ThenInclude(r => r!.ContactPerson)
            .OrderByDescending(d => d.Id)
            .Skip((pageNo - 1) * noOfItems)
            .Take(noOfItems)
            .ToListAsync();

        var creators = await CreatorNames(rows.Select(d => d.CreationBy));

        var data = rows.Select(d => new
        {
            Id = d.Id,
            SupplierName = d.Supplier?.Name ?? string.Empty,
            SupplierId = d.SupplierId ?? 0,
            TransportationVehicleRouteName = d.Route?.NameOfRoute ?? string.Empty,
            TransportationVehicleRouteId = d.RouteId,
            Serial = string.IsNullOrEmpty(d.Serial) ? "0" : d.Serial,
            Date = Fmt.Iso(d.DateOfDeduction),
            CreationDate = Fmt.Iso(d.CreationDate),
            DeductPerRound = d.DeductPerRound,
            Cause = d.Cause ?? string.Empty,
            CreationName = d.CreationBy != null ? creators.GetValueOrDefault(d.CreationBy.Value, string.Empty) : string.Empty,
            CreationBy = d.CreationBy ?? 0,
            SupplierContentPersonIdName = d.Route?.ContactPerson?.Name ?? string.Empty, // frozen typo key (Content vs Contact)
            FromDate = Fmt.Iso(d.FromDate),
            ToDate = Fmt.Iso(d.ToDate),
            typeOfDeduction = d.TypeOfDedct ?? string.Empty, // lowercase frozen key
            // routePrice is the route's CURRENT LineCost (dynamic), not a stored column.
            routePrice = d.Route?.LineCost ?? d.RoutePrice ?? 0, // lowercase frozen key
        });

        return Envelope.SuccessList(data, Envelope.Pagination(pageNo, noOfItems, total));
    }

    /// <summary>GET getAllDeductionExcel — the filtered deductions as a base64 (.xlsx) file.</summary>
    public async Task<object> Excel(DeductionFilters filters)
    {
        var rows = await BuildQuery(filters)
            .Include(d => d.Supplier)
            .Include(d => d.Route).ThenInclude(r => r!.ContactPerson)
            .OrderByDescending(d => d.Id)
            .ToListAsync();

        using var wb = new XLWorkbook();
        var sheet = wb.AddWorksheet("الخصومات");
        sheet.RightToLeft = true;
        string[] headers = { "الخط", "المورد", "السيريال", "التاريخ", "قيمة الخصم", "النوع", "السبب" };
        for (var c = 0; c < headers.Length; c++) sheet.Cell(1, c + 1).Value = headers[c];
        sheet.Row(1).Style.Font.Bold = true;

        var row = 2;
        foreach (var d in rows)
        {
            sheet.Cell(row, 1).Value = d.Route?.NameOfRoute ?? string.Empty;
            sheet.Cell(row, 2).Value = d.Supplier?.Name ?? string.Empty;
            sheet.Cell(row, 3).Value = d.Serial ?? string.Empty;
            sheet.Cell(row, 4).Value = Fmt.IsoDate(d.DateOfDeduction);
            sheet.Cell(row, 5).Value = d.DeductPerRound;
            sheet.Cell(row, 6).Value = (d.TypeOfDedct ?? string.Empty).ToLowerInvariant() == "taxes" ? "ضرائب" : "عادي";
            sheet.Cell(row, 7).Value = d.Cause ?? string.Empty;
            row++;
        }

        return Envelope.Success(Fmt.WorkbookBase64(wb));
    }

    /// <summary>POST AddDeduction — create a deduction. creationBy = caller.</summary>
    public async Task<object> Add(DeductionBody body, int userId)
    {
        var routeId = body.TransportationVehicleRouteId ?? 0;
        if (routeId == 0) return Envelope.Fail("Err101", "TransportationVehicleRouteId is required");
        var entity = new TransportationVehicleRouteDeduction();
        Apply(entity, body, routeId);
        entity.CreationBy = userId;
        _db.TransportationVehicleRouteDeductions.Add(entity);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(entity.Id);
    }

    /// <summary>POST UpdateDeduction — update a deduction. Preserves the original creator.</summary>
    public async Task<object> Update(int id, DeductionBody body)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        var entity = await _db.TransportationVehicleRouteDeductions.FirstOrDefaultAsync(d => d.Id == id);
        if (entity == null) return Envelope.Fail("Err404", "Deduction not found");
        var routeId = body.TransportationVehicleRouteId ?? 0;
        Apply(entity, body, routeId);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }

    private static DateTime? ParseDate(string? s) =>
        string.IsNullOrEmpty(s) ? null : (DateTime.TryParse(s, null, System.Globalization.DateTimeStyles.AdjustToUniversal | System.Globalization.DateTimeStyles.AssumeUniversal, out var d) ? d : null);

    /// <summary>Maps the documented payload to the deduction model columns (no creator).</summary>
    private static void Apply(TransportationVehicleRouteDeduction e, DeductionBody body, int routeId)
    {
        e.SupplierId = body.SupplierId;
        e.RouteId = routeId;
        e.Serial = body.Serial ?? string.Empty;
        e.DateOfDeduction = ParseDate(body.DateOfDeduction) ?? DateTime.UtcNow;
        e.DeductPerRound = body.DeductPerRound ?? 0;
        e.Cause = body.Cause;
        e.TypeOfDedct = body.TypeOfDedct ?? "Normal";
    }
}
